bugwarrior/services/gitlab.py

```diff
@@ -467,8 +568,38 @@ def _get_issue_objs(self, issues, issue_type, repo_map):
             issue_obj.update_extra(extra)
             yield issue_obj
 
+    def _get_todo_objs(self, todos):
+        for project, todo in todos:
+            if project is not None:
+                repo = project
+            else:
+                repo = {
+                    'path': 'the instance',
+                }
+            todo['repo'] = repo['path']
+
+            todo_obj = self.get_issue_for_record(todo)
+            todo_url = todo['target_url']
+            project_name = repo['path']
+            if self.config.project_owner_prefix:
+                project_name = repo['namespace']['path'] + "." + project_name
+            extra = {
+                'issue_url': todo_url,
+                'project': project_name,
+                'namespace': "todo",
+                'type': 'todo',
+                'annotations': [],
+            }
+            todo_obj.update_extra(extra)
+            yield todo_obj
+
     def include(self, issue):
         """ Return true if the issue in question should be included """
+        self.update_repo_map([issue])
```

Comment:
I'll have a  look at that again, maybe tonight.

Reply:
The thing that's giving me a headache is that with the two possible paths we want to filter for `include_repos` and such at two very different places.

* For the project API we want to filter the repositories **before** we query any issues, as otherwise we would be querying issues for projects that we are actually not interested in. As each individual query is very time/intensive we want to reduce the number of necessary calls as much as possible.
* For the issues API we want to first fetch all matching issues and then filter out the projects as we cannot add a repo filter to the issues query.

4f4d188309161ae05033093a870c182facfb3597 and 8394db1c569c86509a945563bdcee3fe9ae64d54 explain my motivation for the way I've implemented this.

Inside `_get_issue_objects` we need the `repo_map` to be filled, as we fetch a lot of info from the project there. The call tp `update_repo_map` in there ensures that the repo information is actually present. The way `update_repo_map` is implemented, it basically works as a caching mechanism. If repo information about the repo with the given idea is already there, then don't do anything. Therefore, calling it twice is not necessarily a bad thing performance-wise.

Moving the `update_repo_map` call out of the functions would effectively add a precondition to the functions, which is not the best thing to introduce. I've added an alternative suggestion in 8ad81b7, but I am not sure whether this approach really carries a benefit over the one discussed here.

Another alternative I could think of is wrapping the `repo_map` into an own cache data structure with something like `get_repo_info_cached` that would either query the info or read it from the cache. This would make the `update_repo_map` function obsolete. Algorithmically not too much would change to my initial approach, but it would be clearer to other developers, that the `repo_map` is acting as a cache. It would not introduce any preconditions to the `include` function or the `_get_issue_objs` function.

Reply:
Very nice solution. This makes the repo map a lot easier to understand.